### ISSUE TITLE
Optimize `bool` and constant embedding

### ISSUE THREAD

**Issue by vlasin** [zkFold/zkfold-base#65]
- [x] Optimize `bool` arithmetization to three constraints
- [ ] Embed constants without creating a constraint

**Comment by echatav** [comment]
With regard to constant embedding, I verified that regardless of the integer chosen at random, embedding creates 1 variable and 1 constraint:
```
Example: constant 6


Compiling the script...

Number of constraints: 1
Number of variables: 1
Script saved: compiled_scripts/constant.json



zkfold-base> Test suite zkfold-base-examples passed
Completed 2 action(s).
hephaestus@Eitans-MacBook-Pro zkfold-base % stack build zkfold-base:zkfold-base-examples
zkfold-base> test (suite: zkfold-base-examples)


Example: constant 93


Compiling the script...

Number of constraints: 1
Number of variables: 1
Script saved: compiled_scripts/constant.json
```
The code for `fromConstant` calls `embed` which is defined as:

```Haskell
embed :: Arithmetic a => a -> ArithmeticCircuit a
embed x = circuit $ newAssigned $ const (fromConstant x)
```

Finally, `newAssigned` is defined as:
```Haskell
    newAssigned :: MonadBluePrint i a m => ClosedPoly i a -> m i
    newAssigned p = newConstrained (\x i -> p x - x i) p
```

From here, it's evident where the 1 constraint comes from. And it's not clear to me how to correctly excise this constraint, assuming that's possible.

**Comment by echatav** [comment]
On the other hand, it's not clear that the definition of `bool` is even calling `fromConstant`. It doesn't appear to be:

```Haskell
instance SymbolicData a x => Conditional (Bool (ArithmeticCircuit a)) x where
    bool brFalse brTrue (Bool b) =
        let f' = pieces brFalse
            t' = pieces brTrue
        in restore $ zipWith (\f t -> b * (t - f) + f) f' t'
```

I must ask, on what basis are we convinced that `bool` can be arithmetized with 2 constaints, not 3?

**Comment by echatav** [comment]
We should also note that the number of variables and constraints for `bool` is highly dependent on the symbolic datatype it operates on:
```
Number of constraints: 0
Number of variables: 0
Script saved: compiled_scripts/conditional0.json

Example: conditional


Compiling the script...

Number of constraints: 3
Number of variables: 4
Script saved: compiled_scripts/conditional.json

Example: conditional on pair


Compiling the script...

Number of constraints: 4
Number of variables: 5
Script saved: compiled_scripts/conditional2.json
```

**Comment by vlasin** [comment]
Closing this issue as constraint-less constant embedding will be solved by #177.

**Comment by TurtlePU** [comment]
On the topic of constant embedding, I would like it to look like this:
```haskell
class (Monad m, FromConstant a v) => MonadCircuit v a m | m -> a, m -> v where
  newConstrained :: ... 
  newRanged :: ...
  constraint :: ...
  newAssigned = ...
```
This way, how to get constant embedding is clear from the interface.

For pure instance, this is guaranteed by `FromConstant a a`; for circuit instance, this is guaranteed by a separate "variable" type which has a case for constants as done in #177.
